# Stack Overflow Survey Analysis:
# Insights into Work Experience, Remote Work, and Compensation

In [2]:
import pandas as pd

In [54]:
df = pd.read_csv('survey_results_schema.csv')

In [4]:
df2 =pd.read_csv('survey_results_public.csv')

In [5]:
print(df.index)
print(df2.index)

RangeIndex(start=0, stop=87, step=1)
RangeIndex(start=0, stop=65437, step=1)


In [6]:
print(df.shape)
print(df2.shape)

(87, 6)
(65437, 114)


In [7]:
print(df.dtypes)
print(df2.dtypes)

qid           object
qname         object
question      object
force_resp    object
type          object
selector      object
dtype: object
ResponseId               int64
MainBranch              object
Age                     object
Employment              object
RemoteWork              object
                        ...   
JobSatPoints_11        float64
SurveyLength            object
SurveyEase              object
ConvertedCompYearly    float64
JobSat                 float64
Length: 114, dtype: object


# 1. Soru:
#### Ankete kaç kişi katıldı?

- Anketi cevplayan herkese bir ID verildiği düşünülürse, ResponseID sütun elemanlarını saymak ankaete katılan kişi sayısını verir.
- Aynı zamanda DataFrame'in satır sayısını 'shape' kodu ile hesaplayarak, kaç tane anket yanıtı olduğunu da görebiliriz.

In [9]:
katilimci_sayisi1 = df2['ResponseId'].count()
print(f"Ankete {katilimci_sayisi1} kişi katılmıştır.")

Ankete 65437 kişi katılmıştır.


In [10]:
katilimci_sayisi2 = df2.shape[0]
print(f"Ankete {katilimci_sayisi2} kişi katılmıştır.")

Ankete 65437 kişi katılmıştır.


# 2. Soru:

#### Kaç kişi tüm soruları yanıtladı?


- Survey_results_schema.csv, anketin yapısını açıklarken survey_results_public.csv, gerçek anket yanıtlarını içeriyor.
- Survey_results_schema.csv içindeki qname sütunu ile survey_results_public.csv içindeki sütun isimleri birebir eşleşiyor.
- Survey_results_schema.csv dosyasında 87 satır var, demek ki ankette sorulan gerçek soru sayısı 87.
- Survey_results_public.csv dosyasında 114 sütun var, yani fazladan 114 - 87 = 27 sistem bilgisi içeren sütun var.
- Tüm soruları yanıtlayanları bulmak için, sadece 87 soru sütununa odaklanıp eksik veri olup olmadığını kontrol ettim.

In [12]:
## Schema Veri Seti'nden geçerli qname satırlarını aldım.

qname_columns = df['qname'].dropna()
qname_columns

0          MainBranch
1                 Age
2          Employment
3          RemoteWork
4               Check
           ...       
82     JobSatPoints_7
83     JobSatPoints_8
84     JobSatPoints_9
85    JobSatPoints_10
86    JobSatPoints_11
Name: qname, Length: 87, dtype: object

In [13]:
##  Public'de mevcut olan ve Shema ile eşleşen sütunları seçtim.

valid_columns = [x for x in qname_columns if x in df2.columns]

# NaN değerlerini, geçerli anket sorularının olduğu sütunlarda kontrol edip, satırları getirmesini istedim.

complete_count = df2.dropna(subset=valid_columns).shape[0]
print(f"Tüm soruları yanıtlayan katılımcı sayısı: {complete_count}")

Tüm soruları yanıtlayan katılımcı sayısı: 6306


# 3. Soru

#### Katılımcıların çalışma deneyimleriyle (WorkExp) ilgili merkezi eğilim ölçüleri nelerdir?

- Ortalama çalışma süresi 11,5 yıl.
- Buna karşılık median 9 yıl. Bu demek oluyor ki verisetinde çok uç değerler bulunmuyor.Ayrıca daha fazla düşük çalışma deneyimi olan katılımcı var.
- Katılımcıların bir çoğunun 3 yıl deneyime sahip olması(mode), katılımcıların çalışma deneyimlerinin kısa süreli olduğunu gösteriyor
- Standart sapma değeri 9,16 olarak büyük bir değerdir.Bu da, verilerin yayılmasının geniş olduğunu ve deneyim seviyelerinin büyük ölçüde farklılık gösterdiğini işaret ediyor.

In [15]:
mean_exp = df2['WorkExp'].mean() 
mean_exp

11.46695663901814

In [16]:
median_exp = df2['WorkExp'].median() 
median_exp

9.0

In [17]:
mode_exp = df2['WorkExp'].mode()[0]
mode_exp

3.0

In [18]:
std_dev_workexp = df2['WorkExp'].std()
std_dev_workexp

9.168708536230668

# 4. Soru

#### Kaç kişi uzaktan çalışıyor?


- Sadece 'Remote' olarak anketi ceavplamış çalışan sayısını hesapladım.
- Sum() ile Remote koşulunu True döndüren katılımcıların sayısını topladım.

In [20]:
remote_workers_count = (df2['RemoteWork'] == 'Remote').sum()
print(f"Uzaktan çalışan katılımcı sayısı: {remote_workers_count}")

Uzaktan çalışan katılımcı sayısı: 20831


# 5. Soru

#### Katılımcıların yüzde kaçı Python ile programlama yapıyor?

* Public veri setindeki sütunları 'df2.columns.tolist()' kodu tararayarak 'LanguageHaveWorkedWith' kolonunu buldum.

In [22]:
## - Verisetinin ilgili sütunun içinde string bir ifade olan 'Python'ı aratıp, na: olan değerlesin false olarak sayılmasını sağlayan,
# True gelen tüm satırları sayan bir kod yazdım.Toplam 'Python' kullanıcı sayısı 30795.

python_users = python_users = df2[df2['LanguageHaveWorkedWith'].str.contains('Python', na=False)].shape[0]
python_users

30795

In [23]:
# Python kullanıcılarını, 1. soruda hesapladığım tüm katılımcılara bölerek yüzdesini aldım.

percentage_python = (python_users / katilimci_sayisi1) * 100
print(f"Katılımcıların yüzde {percentage_python:.2f}'si Python ile programlama yapıyor.")

Katılımcıların yüzde 47.06'si Python ile programlama yapıyor.


# 6. Soru

#### Kaç kişi online kurslar ile programlama öğrendi?


In [25]:
# 'LearnCode' sütununda yer alan online courseları büyük/küçük harf hassasiyeti olmadan, None değerleri kapsamadan topladım.

online_learner = df2['LearnCode'].str.contains('Online Courses or Certification', case=False, na=False).sum()
print(f"Online kurslar ile programlama öğrenen katılımcı sayısı: {online_learner} dir.")

Online kurslar ile programlama öğrenen katılımcı sayısı: 30271 dir.


# 7. Soru

#### Ülkelere göre gruplanmış ve Python ile programlama yapan katılımcılar arasında, her ülkede ortalama ve medyan yıllık tazminat (ConvertedCompYearly) ne kadardır?


In [27]:
#groupby() metodunu kullanmak için python_users değişkenimi tekrar hesaplayarak DataFrame atadım.

python_usersdf = df2[df2['LanguageHaveWorkedWith'].str.contains('Python', na=False)]

In [28]:
# Ülkeye göre gruplayarak ortalama ve medyan yıllık tazminatı hesapladım.


salary_stats_by_country = python_usersdf.groupby('Country')['ConvertedCompYearly'].agg(['mean', 'median'])

# 8. Soru

#### En yüksek tazminata sahip 5 katılımcının eğitim seviyeleri nelerdir?


In [30]:
#ConvertedCompYearly sütunundaki en büyük 5 değeri nlargest fonksiyonu ile seçtim.

top_5_earners = df2.nlargest(5, 'ConvertedCompYearly')
top_5_earners

,ResponseId,MainBranch,Age,Employment,RemoteWork,Check,CodingActivities,EdLevel,LearnCode,LearnCodeOnline,...,JobSatPoints_6,JobSatPoints_7,JobSatPoints_8,JobSatPoints_9,JobSatPoints_10,JobSatPoints_11,SurveyLength,SurveyEase,ConvertedCompYearly,JobSat
15837,15838,I am a developer by profession,25-34 years old,"Independent contractor, freelancer, or self-em...",In-person,Apples,Hobby;Contribute to open-source projects;Boots...,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Books / Physical media;Colleague;On the job tr...,Technical documentation;Blogs;Books;Written Tu...,...,NaN,NaN,NaN,NaN,NaN,NaN,Appropriate in length,Easy,16256603.0,NaN
12723,12724,I am a developer by profession,25-34 years old,"Employed, full-time;Independent contractor, fr...","Hybrid (some remote, some in-person)",Apples,Contribute to open-source projects;Professiona...,"Professional degree (JD, MD, Ph.D, Ed.D, etc.)",Books / Physical media;On the job training;Oth...,Blogs;Stack Overflow;Coding sessions (live or ...,...,100.0,100.0,100.0,100.0,100.0,100.0,Appropriate in length,Neither easy nor difficult,13818022.0,10.0
28379,28380,I am a developer by profession,18-24 years old,"Employed, full-time","Hybrid (some remote, some in-person)",Apples,Contribute to open-source projects,"Professional degree (JD, MD, Ph.D, Ed.D, etc.)",Books / Physical media;On the job training;Oth...,Technical documentation;Blogs;Books;Written Tu...,...,100.0,100.0,100.0,100.0,100.0,100.0,Appropriate in length,Neither easy nor difficult,9000000.0,10.0
17593,17594,I am a developer by profession,35-44 years old,"Employed, full-time",Remote,Apples,Hobby;Professional development or self-paced l...,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Books / Physical media;On the job training;Oth...,Technical documentation;Books;Written Tutorial...,...,75.0,40.0,30.0,40.0,30.0,70.0,Appropriate in length,Easy,6340564.0,6.0
17672,17673,I am a developer by profession,25-34 years old,"Employed, full-time",Remote,Apples,Hobby;School or academic work,"Professional degree (JD, MD, Ph.D, Ed.D, etc.)","Other online resources (e.g., videos, blogs, f...",Technical documentation;Blogs;Stack Overflow;V...,...,25.0,35.0,20.0,10.0,0.0,0.0,Too long,Neither easy nor difficult,4936778.0,7.0


In [31]:
# En yüksek tazminata sahp kişilerin eğitim seviyelerini alan bir kod yazdım.

top_5_education_levels = top_5_earners[['ConvertedCompYearly', 'EdLevel']]
top_5_education_levels

,ConvertedCompYearly,EdLevel
15837,16256603.0,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)"
12723,13818022.0,"Professional degree (JD, MD, Ph.D, Ed.D, etc.)"
28379,9000000.0,"Professional degree (JD, MD, Ph.D, Ed.D, etc.)"
17593,6340564.0,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)"
17672,4936778.0,"Professional degree (JD, MD, Ph.D, Ed.D, etc.)"
